### PR TITLE
Add flush api

### PR DESCRIPTION
**Pull request by aknishid**
### Description:
Flush API to delete all vectors has been added.

### Versions:

- Go Version: 1.19.4
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.8

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/main/CONTRIBUTING.md) document and completed [our CLA agreement](https://cla-assistant.io/vdaas/vald).
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?

### CODE REVIEW

File: example/client/main.go [hunk 1]
```diff
@@ -126,8 +126,8 @@ func main() {
 	glg.Infof("Finish searching %d times", testCount)
 
 	glg.Info("Start removing vector")
-	// Remove indexed 400 vectors from vald cluster.
-	for i := range ids[:insertCount] {
+	// Remove indexed 200 vectors from vald cluster.
+	for i := range ids[:insertCount/2] {
```
Comment (vankichi) [inline example/client/main.go:130]: Could you please define `removeCount` as a constant value like `insertCount`?
And please update the below document.
https://github.com/vdaas/vald/blob/main/docs/tutorial/get-started.md#run-example-code

File: example/client/agent/main.go [hunk 2]
```diff
@@ -140,8 +140,8 @@ func main() {
 	}
 
 	glg.Info("Start removing vector")
-	// Remove indexed 400 vectors from vald cluster.
-	for i := range ids[:insertCount] {
+	// Remove indexed 200 vectors from vald cluster.
+	for i := range ids[:insertCount/2] {
```
Comment (vankichi) [inline example/client/agent/main.go:144]: Could you please define `removeCount` as a constant value like `insertCount`?

File: docs/api/flush.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  ```rpc
+  message Flush {
+      message Request {
+
+      }
+  }
+  ```
+
+  - Flush.Request
+
```
Comment (vankichi) [inline docs/api/flush.md:32]:
```suggestion
  - Flush.Request

    empty
```

File: internal/file/file.go [hunk 4]
```diff
@@ -475,6 +475,43 @@ func ListInDir(path string) ([]string, error) {
 	return files, nil
 }
 
+func DeleteDir(ctx context.Context, path string) (err error) {
+	exists, _, err := ExistsWithDetail(path)
+	if !exists {
+		return err
+	}
```
Comment (vankichi) [inline internal/file/file.go:482]: It seems to the error won't be returned when `exists=true` and `err != nil`.
I'm not sure it is right.

File: pkg/agent/core/ngt/handler/grpc/flush.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/info"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/net/grpc/errdetails"
+	"github.com/vdaas/vald/internal/net/grpc/status"
+	"github.com/vdaas/vald/internal/observability/trace"
+	"go.opentelemetry.io/otel/attribute"
+)
+
+func (s *server) Flush(ctx context.Context, req *payload.Flush_Request) (*payload.Info_Index_Count, error) {
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/flush.go:32]: Could you please add a comment about this method?

File: pkg/agent/core/ngt/handler/grpc/insert.go [hunk 6]
```diff
@@ -85,6 +85,18 @@ func (s *server) Insert(ctx context.Context, req *payload.Insert_Request) (res *
...
+				&errdetails.RequestInfo{
+					RequestId:   req.GetVector().GetId(),
+					ServingData: errdetails.Serialize(req),
+				},
+				&errdetails.ResourceInfo{
+					ResourceType: ngtResourceType + "/ngt.Insert",
+					ResourceName: fmt.Sprintf("%s: %s(%s)", apiName, s.name, s.ip),
+				})
+			log.Warn(err)
+			attrs = trace.StatusCodeAlreadyExists(err.Error())
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/insert.go:99]: I don't know if this status is suitable.
I suppose `aborted` is preferred to `already exists`.

File: pkg/agent/core/ngt/handler/grpc/linear_search.go [hunk 7]
```diff
@@ -86,6 +86,18 @@ func (s *server) LinearSearch(ctx context.Context, req *payload.Search_Request)
...
+				&errdetails.RequestInfo{
+					RequestId:   req.GetConfig().GetRequestId(),
+					ServingData: errdetails.Serialize(req),
+				},
+				&errdetails.ResourceInfo{
+					ResourceType: ngtResourceType + "/ngt.Search",
+					ResourceName: fmt.Sprintf("%s: %s(%s)", apiName, s.name, s.ip),
+				})
+			log.Debug(err)
+			attrs = trace.StatusCodeAborted(err.Error())
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/linear_search.go:100]: you have to update the search API doc cuz it is a new status code.
ref: https://vald.vdaas.org/docs/api/search/#status-code-6

Also, this document has to be updated.
https://vald.vdaas.org/docs/api/status/

File: pkg/agent/core/ngt/handler/grpc/search.go [hunk 8]
```diff
@@ -89,6 +89,18 @@ func (s *server) Search(ctx context.Context, req *payload.Search_Request) (res *
...
+				&errdetails.RequestInfo{
+					RequestId:   req.GetConfig().GetRequestId(),
+					ServingData: errdetails.Serialize(req),
+				},
+				&errdetails.ResourceInfo{
+					ResourceType: ngtResourceType + "/ngt.Search",
+					ResourceName: fmt.Sprintf("%s: %s(%s)", apiName, s.name, s.ip),
+				})
+			log.Debug(err)
+			attrs = trace.StatusCodeAborted(err.Error())
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/search.go:103]: same about the comment on linearsearch.go

File: pkg/agent/core/ngt/service/ngt.go [hunk 9]
```diff
@@ -882,6 +943,74 @@ func (n *ngt) deleteMultiple(uuids []string, now int64, validation bool) (err er
 	return err
 }
 
+func (n *ngt) RegenerateIndexes(ctx context.Context) (err error) {
```
Comment (vankichi) [inline pkg/agent/core/ngt/service/ngt.go:946]: could you please add a comment about this method?

**Review by vankichi — CHANGES_REQUESTED** [review]

File: internal/file/file.go (hunk 4 above)
Comment (aknishid) [reply internal/file/file.go]: Thank you for your feedback. I have made changes to the method to return nil and output a debug log when the path does not exist, in order to ensure that the method exits normally without performing the deletion.

File: pkg/agent/core/ngt/handler/grpc/search.go (hunk 8 above)
Comment (aknishid) [reply pkg/agent/core/ngt/handler/grpc/search.go:103]: I have also added the ABORTED status code to the documentation, following the same approach as in the LinearSearch section. 
Additionally, I have incorporated error handling in the SearchByID method.

**Review by vankichi — CHANGES_REQUESTED** [review]
It looks good for me.
One point, please rebase again and fix the document and proto.

**Review by vankichi — APPROVED** [review]
LGTM

**Review by datelier — DISMISSED** [review]

**Review by kpango — DISMISSED** [review]
LGTM

**Review by datelier — APPROVED** [review]

**Review by vankichi — APPROVED** [review]

**Review by kpango — APPROVED** [review]
LGTM

**Comment by aknishid** [comment]
I've opened a new PR (https://github.com/vdaas/vald/pull/2273), so I will be closing this one.

**Pull request closed without merging**